loki/transform/transform_extract_contained_procedures.py

```diff
@@ -0,0 +1,198 @@
+    # if not already there.
+    arr_shapes = []
+    for shape in (v.shape for v in vars_to_resolve if isinstance(v, Array)):
+        for s in shape:
+            if not isinstance(s, IntLiteral):
+                arr_shapes.append(s)
+    for v in arr_shapes:
+        if not v.name in (var.name for var in vars_to_resolve):
+            vars_to_resolve.append(v)
+
+    # 1. Rescope `vars_to_resolve` into scope of `inner`.
+    # 2. Set the intent of `vars_to_resolve` to "inout" unless specified otherwise in `procedure` scope.
+    vars_to_resolve = [v.rescope(inner) for v in vars_to_resolve]
+    newtypes = map(lambda x: x.type if x.type.intent else x.type.clone(intent = "inout"), vars_to_resolve)
+    vars_to_resolve = tuple(v.clone(type = typ) for v, typ in zip(vars_to_resolve, newtypes))
+
+    ## PRODUCING IMPORTS TO INTRODUCE TO `inner`.
+    # Get all variables from `inner.spec`. Need to search them for resolving kinds and derived types for
+    # variables that do not need resolution.
+    inner_spec_vars = tuple(FindVariables().visit(inner.spec))
+
+    # Produce derived types appearing in `vars_to_resolve` or in `inner.spec` that need to be resolved
+    # from imports of `procedure`.
+    dtype_imports_to_add = tuple(v.type.dtype for v in vars_to_resolve + inner_spec_vars \
+        if isinstance(v.type.dtype, DerivedType))
+
+    # Produce kinds appearing in `vars_to_resolve` or in `inner.spec` that need to be resolved
+    # from imports of `procedure`.
+    kind_imports_to_add = tuple(v.type.kind for v in vars_to_resolve + inner_spec_vars \
+        if v.type.kind and v.type.kind.scope == procedure)
+
+    # Produce all imports to add.
+    # Here the imports are also tidied to only import what is strictly necessary, and with single
+    # USE statements for each module.
+    imports_to_add = []
+    to_lookup_from_imports = dtype_imports_to_add + kind_imports_to_add + var_imports_to_add
+    for val in to_lookup_from_imports:
+        imp = procedure.import_map[val.name]
+        matching_import = tuple(i for i in imports_to_add if i.module == imp.module)
+        if matching_import:
+            # Have already encountered module name, modify existing.
+            matching_import = matching_import[0]
+            imports_to_add.remove(matching_import)
+            newimport = matching_import.clone(symbols = tuple(set(matching_import.symbols + imp.symbols)))
+        else:
+            # Have not encountered the module yet, add new one.
+            newsyms = tuple(s for s in imp.symbols if s.name == val.name)
+            newimport = imp.clone(symbols = newsyms)
+        imports_to_add.append(newimport)
+
+    ## MAKING THE CHANGES TO `inner`
+    # Change `inner` to take `vars_to_resolve` as dummy arguments and add all necessary imports.
+    # After these lines, `inner` should be self-contained or there is a bug.
+    inner.arguments += tuple(vars_to_resolve)
+    inner.variables += tuple(vars_to_resolve)
```

Comment:
This is unnecessary, the arguments update will automatically insert declarations

Reply:
We discussed this, I removed the unnecessary updating of `inner.variables`, but updating `inner.arguments` needs to remain.